File: core/hooks/session-start.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Session Start Hook — 会话启动时自动注入工作流上下文。

借鉴 Trellis session-start.py，在 Claude Code 会话启动时将项目配置、
活跃工作流状态和项目规范注入 AI 上下文。

配置方法（.claude/settings.json）:
    {
      "hooks": {
        "SessionStart": [
          {
            "type": "command",
            "command": "python3 .agents/agent-workflow/hooks/session-start.py"
          }
        ]
      }
    }

输出格式：直接输出文本到 stdout，Claude Code 会将其注入为系统消息。
"""

# Suppress warnings before any other imports
import warnings
warnings.filterwarnings("ignore")

import json
import os
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parents[1] / "utils" / "workflow"
if str(SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPT_DIR))

from path_utils import get_workflow_state_path

# Windows UTF-8 fix (from Trellis)
if sys.platform == "win32":
    import io as _io
    if hasattr(sys.stdout, "reconfigure"):
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    elif hasattr(sys.stdout, "detach"):
        sys.stdout = _io.TextIOWrapper(
            sys.stdout.detach(), encoding="utf-8", errors="replace"
        )


def should_skip() -> bool:
    """Skip injection in non-interactive mode."""
    return os.environ.get("CLAUDE_NON_INTERACTIVE") == "1"


def read_file(path: Path, fallback: str = "") -> str:
    try:
        return path.read_text(encoding="utf-8")
    except (FileNotFoundError, PermissionError, OSError):
        return fallback


def find_project_config(start: Path) -> dict | None:
    """Find and read .claude/config/project-config.json."""
    config_path = start / ".claude" / "config" / "project-config.json"
    if config_path.is_file():
        try:
            return json.loads(config_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            pass
    return None


def find_workflow_state(project_id: str) -> dict | None:
    """Find workflow-state.json for the given project."""
    state_path_raw = get_workflow_state_path(project_id)
    if not state_path_raw:
        return None
    state_path = Path(state_path_raw)
    if state_path.is_file():
        try:
            return json.loads(state_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            pass
    return None


def collect_spec_indices(project_root: Path) -> str:
    """Collect all spec/index.md files."""
    specs_dir = project_root / ".claude" / "specs"
    if not specs_dir.is_dir():
        return ""

    indices = []
    for index_file in sorted(specs_dir.rglob("index.md")):
        rel = index_file.relative_to(project_root)
        content = read_file(index_file)
        if content.strip():
            indices.append(f"### {rel}\n{content[:500]}")

    return "\n\n".join(indices) if indices else ""


def determine_next_action(state: dict | None) -> str:
    """Determine suggested next action based on workflow state."""
    if not state:
        return "没有活跃的工作流。使用 `/workflow start` 开始新任务。"

    status = state.get("status", "idle")
    current_tasks = state.get("current_tasks", [])
    progress = state.get("progress", {})
    completed = progress.get("completed", [])
    failed = progress.get("failed", [])

    if status == "idle":
        return "使用 `/workflow start` 开始新的工作流。"
    elif status == "planned":
        return "规划已完成。使用 `/workflow execute` 开始执行。"
    elif status == "spec_review":
        return "Spec 等待确认。请审查 Spec 文档后确认继续。"
    elif status == "running":
        task_id = current_tasks[0] if current_tasks else "?"
        return f"工作流执行中，当前任务: {task_id}。使用 `/workflow execute` 继续。"
    elif status == "paused":
        return "工作流已暂停。使用 `/workflow execute` 恢复执行。"
    elif status == "failed":
        task_id = current_tasks[0] if current_tasks else "?"
        reason = state.get("failure_reason", "未知")
        return f"任务 {task_id} 失败: {reason}。使用 `/workflow execute --retry` 重试。"
    elif status == "blocked":
        return "工作流被阻塞。使用 `/workflow unblock <dep>` 解除依赖。"
    elif status == "completed":
        return f"工作流已完成 ({len(completed)} 任务)。使用 `/workflow archive` 归档。"
    elif status == "archived":
        return "工作流已归档。使用 `/workflow start` 开始新任务。"
    else:
        return f"当前状态: {status}。使用 `/workflow status` 查看详情。"


def main() -> int:
    if should_skip():
        return 0

    project_root = Path.cwd()

    # 1. Find project config
    config = find_project_config(project_root)
    if not config:
        # No project config — skip injection silently
        return 0

    project = config.get("project") or {}
    project_id = project.get("id") or config.get("projectId", "")
    project_name = project.get("name") or config.get("projectName", project_root.name)

    # 2. Find active workflow state
    state = find_workflow_state(project_id) if project_id else None

    # 3. Collect project specs
    specs = collect_spec_indices(project_root)

    # 4. Build output
    output_parts = []

    output_parts.append(f"<workflow-context>")

    # Project info
    output_parts.append(f"<project-info>")
    output_parts.append(f"项目: {project_name}")
    output_parts.append(f"项目 ID: {project_id}")
    tech = config.get("frameworks", [])
    if tech:
        tech_str = ", ".join(t if isinstance(t, str) else t.get("name", "") for t in tech[:5])
        output_parts.append(f"技术栈: {tech_str}")
    output_parts.append(f"</project-info>")

    # Workflow state
    if state:
        output_parts.append(f"<active-workflow>")
        output_parts.append(f"状态: {state.get('status', 'unknown')}")
        progress = state.get("progress", {})
        completed = progress.get("completed", [])
        total_hint = ""
        tasks_file = state.get("tasks_file", "")
        if tasks_file:
            total_hint = f" (任务文件: {tasks_file})"
        output_parts.append(f"已完成: {len(completed)} 任务{total_hint}")

        current = state.get("current_tasks", [])
        if current:
            output_parts.append(f"当前任务: {', '.join(current)}")

        # Context budget
        metrics = state.get("contextMetrics", {})
        usage = metrics.get("usagePercent", 0)
        if usage > 0:
            output_parts.append(f"上下文使用率: {usage}%")

        output_parts.append(f"</active-workflow>")

    # Next action
    output_parts.append(f"<next-action>")
    output_parts.append(determine_next_action(state))
    output_parts.append(f"</next-action>")

    # Project specs
    if specs:
        output_parts.append(f"<project-specs>")
        output_parts.append(specs)
        output_parts.append(f"</project-specs>")

    # Thinking guides reference
    guides_dir = project_root / ".claude" / "specs" / "guides"
    if guides_dir.is_dir():
        output_parts.append(f"<thinking-guides>")
        output_parts.append("项目包含思维指南，修改代码前请参考:")
        for guide in sorted(guides_dir.glob("*.md")):
            if guide.name != "index.md":
                output_parts.append(f"  - .claude/specs/guides/{guide.name}")
        output_parts.append(f"</thinking-guides>")

    output_parts.append(f"</workflow-context>")

    print("\n".join(output_parts))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: core/hooks/session-start.js
#!/usr/bin/env node
/** @file SessionStart Hook — 会话启动时注入项目配置、工作流状态和护栏提示 */

require('./_utf8')

const fs = require('fs')
const path = require('path')
const { normalizeWindowsShellPath, readProjectConfig } = require('../utils/workflow/path_utils')
const { getStatusMessages } = require('../utils/workflow/workflow_types')
const {
  getWorkflowRuntime,
  getThinkingGuides,
  getCodeSpecsContext,
  collectSpecFiles,
  renderSpecFiles,
  resolveActiveCodeSpecsScope,
} = require('../utils/workflow/task_runtime')
const { shouldSkipInjection } = require('./_skip')

/**
 * 安全读取文件内容
 * @param {string} targetPath - 文件路径
 * @param {string} [fallback=''] - 读取失败时的默认值
 * @returns {string} 文件内容或默认值
 */
function readFile(targetPath, fallback = '') {
  try {
    return fs.readFileSync(targetPath, 'utf8')
  } catch {
    return fallback
  }
}

/**
 * 递归收集 .claude/specs/ 下所有 index.md 的内容摘要
 * @param {string} projectRoot - 项目根目录
 * @returns {string} 拼接后的 spec 索引内容
 */
function collectSpecIndices(projectRoot) {
  const specsDir = path.join(projectRoot, '.claude', 'specs')
  if (!fs.existsSync(specsDir) || !fs.statSync(specsDir).isDirectory()) return ''

  const indices = []
  function walk(currentDir) {
    for (const entry of fs.readdirSync(currentDir, { withFileTypes: true })) {
      const fullPath = path.join(currentDir, entry.name)
      if (entry.isDirectory()) walk(fullPath)
      else if (entry.isFile() && entry.name === 'index.md') {
        const rel = path.relative(projectRoot, fullPath)
        const content = readFile(fullPath)
        if (content.trim()) indices.push(`### ${rel}\n${content.slice(0, 500)}`)
      }
    }
  }
  walk(specsDir)
  return indices.join('\n\n')
}


/**
 * SessionStart hook 主流程：读取项目配置和 workflow 状态，输出上下文与护栏信息
 */
function main() {
  if (shouldSkipInjection()) return
  const projectRoot = path.resolve(normalizeWindowsShellPath(process.cwd()))
  const config = readProjectConfig(projectRoot)
  if (!config) return

  const project = config.project || {}
  const projectId = project.id || config.projectId || ''
  const projectName = project.name || config.projectName || path.basename(projectRoot)
  const runtime = getWorkflowRuntime(projectRoot)
  const state = projectId && runtime.projectId === projectId ? runtime.state : null
  const stateParseError = runtime.stateParseError || null
  const specs = collectSpecIndices(projectRoot)

  const parts = []
  parts.push('<workflow-context>')
  parts.push('<project-info>')
  parts.push(`项目: ${projectName}`)
  parts.push(`项目 ID: ${projectId}`)
  const frameworks = Array.isArray(config.frameworks) ? config.frameworks : []
  if (frameworks.length) {
    const tech = frameworks.slice(0, 5).map((item) => typeof item === 'string' ? item : item.name || '').filter(Boolean).join(', ')
    if (tech) parts.push(`技术栈: ${tech}`)
  }
  parts.push('</project-info>')

  if (stateParseError) {
    parts.push('<workflow-warning>')
    parts.push(`workflow 状态文件解析失败: ${stateParseError}。请检查 workflow-state.json 是否损坏。`)
    parts.push('</workflow-warning>')
  }

  if (state) {
    parts.push('<active-workflow>')
    parts.push(`状态: ${state.status || 'unknown'}`)
    const progress = state.progress || {}
    const completed = progress.completed || []
    const tasksFile = state.plan_file || state.tasks_file || ''
    parts.push(`已完成: ${completed.length} 任务${tasksFile ? ` (任务文件: ${tasksFile})` : ''}`)
    const current = state.current_tasks || []
    if (current.length) parts.push(`当前任务: ${current.join(', ')}`)
    const usage = ((state.contextMetrics || {}).usagePercent) || 0
    if (usage > 0) parts.push(`上下文使用率: ${usage}%`)
    parts.push('</active-workflow>')
  }

  const { nextAction, guardrail } = getStatusMessages(state, { verbose: true })
  parts.push('<next-action>')
  parts.push(nextAction)
  parts.push('</next-action>')

  parts.push('<workflow-guardrail>')
  parts.push(guardrail)
  parts.push('</workflow-guardrail>')

  if (specs) {
    parts.push('<project-specs>')
    parts.push(specs)
    parts.push('</project-specs>')
  }

  const guides = getThinkingGuides(projectRoot)
  if (guides && guides.files.length) {
    parts.push('<thinking-guides>')
    parts.push('项目包含思维指南，修改代码前请参考:')
    for (const guide of guides.files) parts.push(`  - ${guide.displayPath}`)
    if (guides.legacyWarning) parts.push(`兼容提示: ${guides.legacyWarning}`)
    parts.push('</thinking-guides>')
  }

  // v3 Stage B1: paths-only 判定 + runtime.scope 优先级 + scopeDenied 渲染
  // 切 paths-only 的条件：
  //   1) monorepo + codeSpecs.runtime.scope 非 null 且解析出具体 activePackage（或 scopeDenied）
  //   2) 显式 env SPEC_INJECT_MODE=paths-only
  //   3) 估算 collector 产出的文件量 > 预算 70% 阈值
  const scope = resolveActiveCodeSpecsScope(runtime, config, {})
  const runtimeScopeConfigured = !!((config.codeSpecs || {}).runtime || {}).scope
  const projectType = (config.project || {}).type
  const envForcePathsOnly = process.env.SPEC_INJECT_MODE === 'paths-only'
  const sessionMaxChars = 2000
  let codeSpecsBlock = null
  let codeSpecsScopeAttr = 'overview'

  if (scope && scope.scopeDenied) {
    // scopeDenied → 输出空段 + 原因提示，不回退全树
    const denied = collectSpecFiles(projectRoot, scope, {})
    codeSpecsBlock = renderSpecFiles(denied, { mode: 'paths-only' })
    codeSpecsScopeAttr = 'scope-denied'
  } else {
    const collection = collectSpecFiles(projectRoot, scope || null, {})
    if (collection && collection.files.length) {
      // 估算字节：每文件 ~500 字内容 + 路径开销
      const estimated = collection.files.length * 520
      const pathsOnlyByBudget = estimated > sessionMaxChars * 0.7
      const pathsOnlyByRuntimeScope = projectType === 'monorepo'
        && runtimeScopeConfigured
        && scope && scope.activePackage
      const usePathsOnly = envForcePathsOnly || pathsOnlyByRuntimeScope || pathsOnlyByBudget
      if (usePathsOnly) {
        codeSpecsBlock = renderSpecFiles(collection, { mode: 'paths-only' })
        codeSpecsScopeAttr = scope && scope.activePackage ? `paths-only:${scope.activePackage}` : 'paths-only'
      } else {
        codeSpecsBlock = renderSpecFiles(collection, {
          mode: 'digest',
          maxChars: sessionMaxChars,
          rootIndexBudget: 200,
          layerIndexBudget: 120,
        })
      }
    } else if (collection && !collection.files.length && collection.dirInfo) {
      // code-specs 目录存在但空：简短提示
      codeSpecsBlock = '_code-specs 目录存在但无已填充内容（运行 `/spec-update` 开始沉淀）_'
      codeSpecsScopeAttr = 'empty'
    } else {
      // 无 code-specs 目录：兼容历史行为，保持静默
      codeSpecsBlock = getCodeSpecsContext(projectRoot, sessionMaxChars, { rootIndexBudget: 200, layerIndexBudget: 120 })
    }
  }

  if (codeSpecsBlock) {
    parts.push(`<project-code-specs role="advisory" scope="${codeSpecsScopeAttr}">`)
    parts.push(codeSpecsBlock)
    parts.push('</project-code-specs>')
  }

  parts.push('</workflow-context>')

  // first-reply-notice 默认 OFF：在 strict-output 交互场景（用户要求首轮 JSON / patch / commit message）
  // 注入"必须以中文 sentence 开头"会污染输出。用户想可视确认 SessionStart 已跑时，
  // 显式 export AGENT_WORKFLOW_FIRST_REPLY_NOTICE=1 启用。
  if (process.env.AGENT_WORKFLOW_FIRST_REPLY_NOTICE === '1') {
    parts.push('<first-reply-notice>')
    parts.push('On the first visible assistant reply in this session, begin with exactly one short Chinese sentence:')
    parts.push('SessionStart 已注入：workflow / 当前任务 / git / specs。')
    parts.push('Then continue. One-shot: do not repeat after the first reply in the same session.')
    parts.push('</first-reply-notice>')
  }

  process.stdout.write(parts.join('\n'))
}

try {
  main()
} catch {
  process.exitCode = 0
}
